federationapi/routing/send.go

```diff
@@ -164,6 +165,31 @@ func Send(
 	mu *internal.MutexByRoom,
 	servers federationAPI.ServersInRoomProvider,
 ) util.JSONResponse {
+	// First we should check if this origin has already submitted this
+	// txn ID to us. If they have and the txnIDs map contains an entry,
+	// the transaction is still being worked on. The new client can wait
+	// for it to complete rather than creating more work.
+	index := string(request.Origin()) + string(txnID)
+	if v, ok := txnIDs.Load(index); ok {
+		// This origin already submitted this txn ID to us, and the work
+		// is still taking place, so we'll just wait for it to finish.
+		select {
+		case <-httpReq.Context().Done():
+			return util.JSONResponse{Code: http.StatusRequestTimeout}
```

Comment:
Yes, that's right, because if the caller attempts a second time and then also times out, we don't want to parse the body of their second HTTP request — the first one is already in flight.